##CUNY - Data 612

#Project 5
#Implementing a Recommender System on Spark

##Peter Kowalchuk and Violeta Stoyanova

#Introduction

The purpose of this project is to build a recommender system using the same dataset as in previous project https://rpubs.com/pkowalchuk/588534 and compare the performance with the previous iteration. The dataset is derived from Amazon and consists of users' ratings of instant videos. In the previous project, Project 4, both item (IBCF) and user (UBCF) based collaborative filtering recommenders were designed on a reduced set (80k entries) of data from the Amazon dataset. It was not possible to train a model with a larger dataset on an on-premise setting or without using a distributed machine-learning framework.
In Project 5, the whole dataset will be used (over 500k entries) and will be adapted to work with Apache Spark via Databricks Community Edition. An Alternating Least Squares (ALS) model was chosen instead of ICBF and UBCF. ALS is an algorithm very well suited for distributed computing since it is an iterative process. 
“ALS recommender is a matrix factorization algorithm that uses Alternating Least Squares with Weighted-Lamda-Regularization (ALS-WR). It factors the user to item matrix A into the user-to-feature matrix U and the item-to-feature matrix M: It runs the ALS algorithm in a parallel fashion.  The ALS algorithm should uncover the latent factors that explain the observed user to item ratings and tries to find optimal factor weights to minimize the least squares between predicted and actual ratings.” –
https://www.elenacuoco.com/2016/12/22/alternating-least-squares-als-spark-ml/?cn-reloaded=1

#Load Data

Datasets are available on Julian McAuley’s site: http://jmcauley.ucsd.edu/data/amazon/links.html
The csv file for Instant Video was uploaded onto the databricks file system using the UI. The file is then loaded onto a Databricks dataframe.

In [4]:
# File location and type
file_location = "/FileStore/tables/ratings_Amazon_Instant_Video.csv"
file_type = "csv"

# CSV options
infer_schema = "false"
first_row_is_header = "false"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
data = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load(file_location)

display(data)

_c0 _c1 _c2 _c3 A1EE2E3N7PW666 B000GFDAUG 5.0 1202256000 AGZ8SM1BGK3CK B000GFDAUG 5.0 1198195200 A2VHZ21245KBT7 B000GIOPK2 4.0 1215388800 ACX8YW2D5EGP6 B000GIOPK2 4.0 1185840000 A9RNMO9MUSMTJ B000GIOPK2 2.0 1281052800 A3STFVPM8NHJ7B B000GIOPK2 5.0 1203897600 A2582KMXLK2P06 B000GIOPK2 5.0 1205884800 A1TZCLCW9QGGBH B000GIOPK2 4.0 1209427200 A2E2I6B878CRMA B000GIOPK2 5.0 1378684800 AD5MZA8SOVMPJ B000GIOPK2 5.0 1218240000 A3IE1M3QVUKIJN B000GIOPK2 5.0 1251763200 AZ1MUCW76BDL1 B000GIOPK2 5.0 1361145600 A2XNOB1T796Y6B B000GIOPK2 5.0 1233532800 A12DO7F3TT123V B000GIOPK2 2.0 1189987200 A2UN6AL460C8J4 B000GIOPK2 1.0 1391299200 AVYBQU4XX5QR4 B000GIOPK2 4.0 1363046400 AVE3EF44DFS0C B000GIOPK2 5.0 1190937600 A27AWN5G5GT6RP B000GIOPK2 5.0 1328745600 A35KJPLBWHF5GJ B000GIOPK2 5.0 1257033600 ADP0IVFAH8EJF B000GIOPK2 5.0 1357948800 A2MBHFCT26IYQ6 B000GIOPK2 5.0 1235433600 AONBKXBKBR8BO B000GIOPK2 5.0 1379894400 A2K72PJNBUP2TI B000GIOPK2 5.0 1404086400 A36ANEVX17XF9S B000GIOPK2 5.0 1191456000 A3T0DNK02KT55Q B000GIOPK2 5.0 1404518400 A2D18BJF9Z3QUS B000GIOPK2 4.0 1361664000 A7ZYU4SOLTUGB B000GIOPK2 5.0 1237420800 AJKWF4W7QD4NS B000GIOPK2 3.0 1186185600 A17L10630VSW76 B000GIOPK2 4.0 1204675200 A3EP4NETI2AB94 B000GIOPK2 4.0 1194912000 A2GV5HOVHKYDCC B000GIOPK2 5.0 1194134400 A25G7I1L842JX B000GIOPK2 5.0 1217289600 A25TRME1TXF4ZE B000GIOPK2 4.0 1354752000 A30EVNW3J601PN B000GIOPK2 5.0 1358121600 A3AV53EIA4EO41 B000GIOPK2 2.0 1259107200 A05132041L0G3P5ASPHR0 B000GIOPK2 5.0 1353110400 A3DIMK33JFV9ER B000GIOPK2 4.0 1216857600 A28UGF8IFWKK8B B000GIOPK2 5.0 1191110400 A301CDXWK3GBBQ B000GIOPK2 5.0 1190332800 A1ZAVMB4XZL8KA B000GIPKWY 5.0 1289088000 A223A5BICW5T32 B000GIPKWY 4.0 1196121600 A3I8OURC83NM6V B000GIPKWY 5.0 1214352000 A284NRFS8VDI3X B000GIPKWY 1.0 1208822400 A13PPQ2GI1IJYM B000GIPKWY 5.0 1404950400 A38AAPXSJN4C5G B000GIPKWY 4.0 1310256000 A3D8ZJ5Y8Z2GEG B000GIPKWY 5.0 1208908800 AFM9DJYPNM7VT B000GIPKWY 2.0 1351814400 A10KA69V7ITA5S B000GIPKWY 5.0 1231200000 A3RFDAOM1TBQ15 B000GIPKWY 5.0 1217462400 A333G9I2SYHC0Y B000GIPKWY 5.0 1292025600 ALNQS7RJIYG6 B000GIPKWY 2.0 1221782400 A3410YP65KJQ3M B000GIPKWY 5.0 1387497600 A153NZD2WZN5S3 B000GIPKWY 5.0 1273017600 A143UP2GFU795Z B000GIPKWY 1.0 1392595200 A1OBAJ2VJCAV0B B000GIPKWY 4.0 1264982400 A1UUMBYR6F3VHA B000GIPKWY 5.0 1270944000 A2QIYVBS2S0UAQ B000GIPKWY 1.0 1271808000 A2P7ED0A9AVCDU B000GIPKWY 3.0 1282608000 A10VPZUGZAO0GG B000GIPKWY 5.0 1270252800 A3E3ELTAMZ32K3 B000GIPKWY 4.0 1383782400 A21YD7FBMV01HP B000GJUQ7M 1.0 1164931200 A1LMX2N66EYKB4 B000GJUQ7M 5.0 1170374400 AGPOJFLTYSMFW B000GJUQ7M 5.0 1169251200 A35O4YSNIEHCR B000GK0NBK 5.0 1157932800 A1ALC8JP5D7C3C B000GK0NBK 2.0 1173139200 A1UU0WHNOXGGXB B000GK51HG 5.0 1200528000 AVZ30RTKJHKKT B000GK6NFK 1.0 1229817600 AP7JOAYYKOGVW B000GK6NFK 5.0 1221868800 A21G21TFGZNBUR B000GK6NFK 5.0 1394928000 ARNNW9J2DG005 B000GK6NFK 5.0 1244851200 A22U1ZY8IMNS7C B000GK6NFK 5.0 1355356800 A21Q8TYJO50OCU B000GK6NFK 5.0 1365465600 A1BWNF5GHHI9SI B000GK6NFK 1.0 1311984000 AHPQ16I8JWH6L B000GK6NFK 3.0 1344816000 AXYESYR1GNGM0 B000GK6NFK 5.0 1251244800 AWRQ9YBI64ZJB B000GK6NFK 5.0 1343952000 A2JP8K38NLHBQJ B000GK6NFK 5.0 1296691200 A16UW55IAXI9LW B000GK6NFK 5.0 1377043200 A55KA3FEBNDL B000GK6NFK 5.0 1245456000 A3QW44Q3C856NR B000GK6NFK 5.0 1290470400 A7SH32KDX7BS2 B000GK6NFK 5.0 1391731200 A2VRO53CM5EP28 B000GK6NFK 5.0 1358553600 A1CU3ANLIPE538 B000GK6NFK 5.0 1367884800 A3VI7N9O6K3MAF B000GK6NFK 5.0 1381536000 A3P63XV3B7IA36 B000GK6NFK 5.0 1381104000 A28SXDX87WNBDZ B000GK6NFK 5.0 1394928000 A1XLZA0EKRUDJR B000GK6NFK 5.0 1384819200 A3VOYWGL8SZ6T8 B000GK6NFK 5.0 1368230400 A12DJ1784JEUOL B000GK6NFK 4.0 1227484800 AV6AFHG3U0U9A B000GK6NFK 4.0 1242777600 ABUZCXVWZLHC9 B000GK6NFK 3.0 1223683200 A1DT8VXDWGY02B B000GK6NFK 4.0 1401667200 A3KTYXYEWMJHUH B000GK6NFK 5.0 1378944000 A2FVDZ9F0YPIFP B000GK6NFK 5.0 1384732800 A1SYZI8XAFW9HD B000GK6NFK 4.0 1227571200 A1H5G60239IB8D B000GK6NFK 2.0 1367366400 A2P6REPFF51MTE B000GK6NFK 5.0 1283126400 A24YV7DUO05ATU B000GK6NFK 5.

Drop timestamp from dataset and rename columns

In [6]:
data=data.selectExpr("_c0 as userId","_c1 as itemId","_c2 as rating")
display(data)

userId itemId rating A1EE2E3N7PW666 B000GFDAUG 5.0 AGZ8SM1BGK3CK B000GFDAUG 5.0 A2VHZ21245KBT7 B000GIOPK2 4.0 ACX8YW2D5EGP6 B000GIOPK2 4.0 A9RNMO9MUSMTJ B000GIOPK2 2.0 A3STFVPM8NHJ7B B000GIOPK2 5.0 A2582KMXLK2P06 B000GIOPK2 5.0 A1TZCLCW9QGGBH B000GIOPK2 4.0 A2E2I6B878CRMA B000GIOPK2 5.0 AD5MZA8SOVMPJ B000GIOPK2 5.0 A3IE1M3QVUKIJN B000GIOPK2 5.0 AZ1MUCW76BDL1 B000GIOPK2 5.0 A2XNOB1T796Y6B B000GIOPK2 5.0 A12DO7F3TT123V B000GIOPK2 2.0 A2UN6AL460C8J4 B000GIOPK2 1.0 AVYBQU4XX5QR4 B000GIOPK2 4.0 AVE3EF44DFS0C B000GIOPK2 5.0 A27AWN5G5GT6RP B000GIOPK2 5.0 A35KJPLBWHF5GJ B000GIOPK2 5.0 ADP0IVFAH8EJF B000GIOPK2 5.0 A2MBHFCT26IYQ6 B000GIOPK2 5.0 AONBKXBKBR8BO B000GIOPK2 5.0 A2K72PJNBUP2TI B000GIOPK2 5.0 A36ANEVX17XF9S B000GIOPK2 5.0 A3T0DNK02KT55Q B000GIOPK2 5.0 A2D18BJF9Z3QUS B000GIOPK2 4.0 A7ZYU4SOLTUGB B000GIOPK2 5.0 AJKWF4W7QD4NS B000GIOPK2 3.0 A17L10630VSW76 B000GIOPK2 4.0 A3EP4NETI2AB94 B000GIOPK2 4.0 A2GV5HOVHKYDCC B000GIOPK2 5.0 A25G7I1L842JX B000GIOPK2 5.0 A25TRME1TXF4ZE B000GIOPK2 4.0 A30EVNW3J601PN B000GIOPK2 5.0 A3AV53EIA4EO41 B000GIOPK2 2.0 A05132041L0G3P5ASPHR0 B000GIOPK2 5.0 A3DIMK33JFV9ER B000GIOPK2 4.0 A28UGF8IFWKK8B B000GIOPK2 5.0 A301CDXWK3GBBQ B000GIOPK2 5.0 A1ZAVMB4XZL8KA B000GIPKWY 5.0 A223A5BICW5T32 B000GIPKWY 4.0 A3I8OURC83NM6V B000GIPKWY 5.0 A284NRFS8VDI3X B000GIPKWY 1.0 A13PPQ2GI1IJYM B000GIPKWY 5.0 A38AAPXSJN4C5G B000GIPKWY 4.0 A3D8ZJ5Y8Z2GEG B000GIPKWY 5.0 AFM9DJYPNM7VT B000GIPKWY 2.0 A10KA69V7ITA5S B000GIPKWY 5.0 A3RFDAOM1TBQ15 B000GIPKWY 5.0 A333G9I2SYHC0Y B000GIPKWY 5.0 ALNQS7RJIYG6 B000GIPKWY 2.0 A3410YP65KJQ3M B000GIPKWY 5.0 A153NZD2WZN5S3 B000GIPKWY 5.0 A143UP2GFU795Z B000GIPKWY 1.0 A1OBAJ2VJCAV0B B000GIPKWY 4.0 A1UUMBYR6F3VHA B000GIPKWY 5.0 A2QIYVBS2S0UAQ B000GIPKWY 1.0 A2P7ED0A9AVCDU B000GIPKWY 3.0 A10VPZUGZAO0GG B000GIPKWY 5.0 A3E3ELTAMZ32K3 B000GIPKWY 4.0 A21YD7FBMV01HP B000GJUQ7M 1.0 A1LMX2N66EYKB4 B000GJUQ7M 5.0 AGPOJFLTYSMFW B000GJUQ7M 5.0 A35O4YSNIEHCR B000GK0NBK 5.0 A1ALC8JP5D7C3C B000GK0NBK 2.0 A1UU0WHNOXGGXB B000GK51HG 5.0 AVZ30RTKJHKKT B000GK6NFK 1.0 AP7JOAYYKOGVW B000GK6NFK 5.0 A21G21TFGZNBUR B000GK6NFK 5.0 ARNNW9J2DG005 B000GK6NFK 5.0 A22U1ZY8IMNS7C B000GK6NFK 5.0 A21Q8TYJO50OCU B000GK6NFK 5.0 A1BWNF5GHHI9SI B000GK6NFK 1.0 AHPQ16I8JWH6L B000GK6NFK 3.0 AXYESYR1GNGM0 B000GK6NFK 5.0 AWRQ9YBI64ZJB B000GK6NFK 5.0 A2JP8K38NLHBQJ B000GK6NFK 5.0 A16UW55IAXI9LW B000GK6NFK 5.0 A55KA3FEBNDL B000GK6NFK 5.0 A3QW44Q3C856NR B000GK6NFK 5.0 A7SH32KDX7BS2 B000GK6NFK 5.0 A2VRO53CM5EP28 B000GK6NFK 5.0 A1CU3ANLIPE538 B000GK6NFK 5.0 A3VI7N9O6K3MAF B000GK6NFK 5.0 A3P63XV3B7IA36 B000GK6NFK 5.0 A28SXDX87WNBDZ B000GK6NFK 5.0 A1XLZA0EKRUDJR B000GK6NFK 5.0 A3VOYWGL8SZ6T8 B000GK6NFK 5.0 A12DJ1784JEUOL B000GK6NFK 4.0 AV6AFHG3U0U9A B000GK6NFK 4.0 ABUZCXVWZLHC9 B000GK6NFK 3.0 A1DT8VXDWGY02B B000GK6NFK 4.0 A3KTYXYEWMJHUH B000GK6NFK 5.0 A2FVDZ9F0YPIFP B000GK6NFK 5.0 A1SYZI8XAFW9HD B000GK6NFK 4.0 A1H5G60239IB8D B000GK6NFK 2.0 A2P6REPFF51MTE B000GK6NFK 5.0 A24YV7DUO05ATU B000GK6NFK 5.0 A24WXVWTRPC6J3 B000GK6NFK 5.0 A3OGUQOD70EMMK B000GK6NFK 4.0 A3MTD1X2STS73G B000GK6NFK 4.0 A2CN3RAZ6YMG7B B000GK6NFK 5.0 A1LFPMXA63SC8N B000GK6NFK 5.0 A3FF40QHATHVK0 B000GK6NFK 5.0 ASLA77MR9QKGP B000GK6NFK 2.0 A1FTD75LGWVKPH B000GK6NFK 5.0 A12Y211FLZBU62 B000GK6NFK 5.0 A1QN643GXIE1Q5 B000GK6NFK 5.0 A2L513SLW6IVGO B000GK6NFK 5.0 A3H8F10BFSZ14U B000GK6NFK 5.0 A1KNGTJWOW8NA9 B000GK6NFK 1.0 ADQESAUCAQFUS B000GK6NFK 5.0 ARYRX0J70PLZQ B000GK6NFK 5.0 A1N2C8M6DQYIEW B000GK6NFK 5.0 AA738P7M3ZKAY B000GK6NFK 5.0 A397ISZD1AW5YM B000GK6NFK 1.0 AJ0PB8EHFXLM5 B000GK6NFK 4.0 A30TYK11RSXNPE B000GK6NFK 4.0 A21NJ6Q8H9A5HH B000GK6NFK 5.0 A3LKO1N1W6UOEP B000GK6NFK 3.0 A5BHKRNK8WGSE B000GK6NFK 5.0 A18CMGIQZ1OAA2 B000GK6NFK 5.0 A2OENV0Q6XM7RV B000GK6NFK 5.0 ALYYCICBKYSQ7 B000GK6NFK 5.0 A22T5BBSN1RKNU B000GK6NFK 5.0 A1GCZ7WDSXRIT0 B000GK7DPY 5.0 A1QKECGNKWOYX5 B000GK7DPY 1.0 AGPOJFLTYSMFW B000GK7DPY 5.0 A2GD669PGTMJZ B000GOTJGG 1.0 A1NZLRAZJGD99W B000GOTJGG 2.0 A1P9VRIY3R9YTT B000GOTJGG 4.0 A1DQQO0A82SCKO B000GOV10S 5.0 A23BUI9NSDYUTY B000GOW7RE 5.0 A3G8Y1R1MEBZ6L B000

In [7]:
print("Total number of user/item ratings in dataset"),str(data.count())

Total number of user/item ratings in dataset
 Out[ 3 ]: (None, '583933')

Build dataframe with average ratings for movies and movie counts

In [9]:
from pyspark.sql import functions as F

item_ids_with_avg_ratings_df = data.groupBy('itemId').agg(F.count(data.rating).alias("Count"), F.avg(data.rating).alias("average"))
print ('item_ids_with_avg_ratings_df:')
item_ids_with_avg_ratings_df.show(5, truncate=False)

item_ids_with_avg_ratings_df:
+----------+-----+-----------------+
itemId |Count|average |
+----------+-----+-----------------+
B000OC3FZQ|78 |4.551282051282051|
B000P41FAA|2 |3.5 |
B000RKQEQW|2 |5.0 |
B000TS73MG|2 |5.0 |
B000U5IH7I|12 |4.583333333333333|
+----------+-----+-----------------+
only showing top 5 rows

In [10]:
print("Total number of items in dataset"),item_ids_with_avg_ratings_df.count()

Total number of items in dataset
 Out[ 5 ]: (None, 23965)

#Select relevant data

To build this model we are going to select items with a 1000 or more reviews

In [12]:
from pyspark.sql.functions import col

In [13]:
items_with_many_ratings_or_more = item_ids_with_avg_ratings_df.filter(item_ids_with_avg_ratings_df.Count>1000).sort(col("average").desc())
print ('Items with highest ratings:')
items_with_many_ratings_or_more.show(20, truncate=False)

Items with highest ratings:
+----------+-----+------------------+
itemId |Count|average |
+----------+-----+------------------+
B009IG59DK|1309 |4.881588999236058 |
B00CE10BD4|1067 |4.878163074039363 |
B00DTOYIIE|6927 |4.857947163274145 |
B003NRWVMC|1132 |4.845406360424028 |
B0083IJGBU|1401 |4.844396859386153 |
B009IF7092|1191 |4.812762384550798 |
B00APE00H4|4501 |4.808487002888247 |
B004MWZLYC|3286 |4.804930006086427 |
B0040HNZTW|2136 |4.7911985018726595|
B00FMO8DXU|1694 |4.78689492325856 |
B0099JKR6U|3884 |4.78681771369722 |
B00D1N6OQ2|1005 |4.782089552238806 |
B006Z48TZS|3680 |4.775271739130435 |
B004KPX56C|1303 |4.768994627782042 |
B008QTTGGG|3230 |4.768730650154799 |
B008BQ8VOC|2723 |4.764965112008814 |
B009FZFONO|6838 |4.763819830359754 |
B005544TRQ|4540 |4.754405286343612 |
B00F406S2U|1172 |4.739761092150171 |
B00BLCHL4Y|3596 |4.735261401557286 |
+----------+-----+------------------+
only showing top 20 rows

In [14]:
print("Number of items with many ratings:"),items_with_many_ratings_or_more.count()

Number of items with many ratings:
 Out[ 8 ]: (None, 69)

In [15]:
df=data.join(items_with_many_ratings_or_more,data.itemId==items_with_many_ratings_or_more.itemId).drop(items_with_many_ratings_or_more.itemId)
print("Number of user/item ratings with items with more than 1000 ratings"),df.count()

Number of user/item ratings with items with more than 1000 ratings
 Out[ 9 ]: (None, 192377)

In [16]:
display(df)

userId rating itemId Count average A3OSZISQRMTWZX 5.0 B003NRWVMC 1132 4.845406360424028 AUMYZ3TS44MNP 5.0 B003NRWVMC 1132 4.845406360424028 A3RA3ZTC0HDP9U 5.0 B003NRWVMC 1132 4.845406360424028 A18NIX1QH8L1Z9 4.0 B003NRWVMC 1132 4.845406360424028 A1HQ4RXXJNS5TC 5.0 B003NRWVMC 1132 4.845406360424028 A36GS5K3YYXFMX 5.0 B003NRWVMC 1132 4.845406360424028 A34IPQQL7NIJQL 5.0 B003NRWVMC 1132 4.845406360424028 A20ZO8UFW5EFS6 5.0 B003NRWVMC 1132 4.845406360424028 A38IOP4TW32Y2Q 5.0 B003NRWVMC 1132 4.845406360424028 A133G8UNAV44BP 5.0 B003NRWVMC 1132 4.845406360424028 AEKZL04PZT00E 5.0 B003NRWVMC 1132 4.845406360424028 A2ZTOTUEBOOFQO 4.0 B003NRWVMC 1132 4.845406360424028 A1ZN17SY8B3LCZ 5.0 B003NRWVMC 1132 4.845406360424028 A3P6L56GUQ85Q2 5.0 B003NRWVMC 1132 4.845406360424028 A1AUP6RUJD01IW 5.0 B003NRWVMC 1132 4.845406360424028 A1S7DJPSFDSEIW 5.0 B003NRWVMC 1132 4.845406360424028 AFTCML5ZKDJ2F 5.0 B003NRWVMC 1132 4.845406360424028 A1EUIHORKSTT3T 5.0 B003NRWVMC 1132 4.845406360424028 AP3JFZYTXTD68 5.0 B003NRWVMC 1132 4.845406360424028 AT0CF2RA9EZHG 5.0 B003NRWVMC 1132 4.845406360424028 A1M0OOSTWL760H 5.0 B003NRWVMC 1132 4.845406360424028 A11J7EOZFK4VX0 5.0 B003NRWVMC 1132 4.845406360424028 A3MICL5RG8QZX7 5.0 B003NRWVMC 1132 4.845406360424028 A2CUZL460CVQP5 5.0 B003NRWVMC 1132 4.845406360424028 A1GP6FF2KAJ4OQ 5.0 B003NRWVMC 1132 4.845406360424028 A4YOGHJ318UGY 4.0 B003NRWVMC 1132 4.845406360424028 AYJKFTHNF6BYD 5.0 B003NRWVMC 1132 4.845406360424028 A2MYD85NGIUWE6 5.0 B003NRWVMC 1132 4.845406360424028 APT5R9GEC52L8 5.0 B003NRWVMC 1132 4.845406360424028 A5TDOCTCGY19V 5.0 B003NRWVMC 1132 4.845406360424028 A3NT2KWVZ8WEBJ 5.0 B003NRWVMC 1132 4.845406360424028 A2QYY2GU66WDXU 5.0 B003NRWVMC 1132 4.845406360424028 AGOYTDOB23CNY 5.0 B003NRWVMC 1132 4.845406360424028 A1C7AQVHUDWFY4 4.0 B003NRWVMC 1132 4.845406360424028 ASV9ERSOWJN5V 5.0 B003NRWVMC 1132 4.845406360424028 A2G10P9DQS555B 5.0 B003NRWVMC 1132 4.845406360424028 ALE8NCB09O7PG 5.0 B003NRWVMC 1132 4.845406360424028 A193F6OTQGTWFJ 5.0 B003NRWVMC 1132 4.845406360424028 A2VDQ2WIVGFNFD 5.0 B003NRWVMC 1132 4.845406360424028 A3MEUXPK3NVFK7 5.0 B003NRWVMC 1132 4.845406360424028 A1LJK5FVKFDXZU 4.0 B003NRWVMC 1132 4.845406360424028 AIDK7FMI74XI 5.0 B003NRWVMC 1132 4.845406360424028 AWXT84BCMQ5QK 5.0 B003NRWVMC 1132 4.845406360424028 A1U58K0US2NHWV 5.0 B003NRWVMC 1132 4.845406360424028 A2AVVEH1V53IJW 5.0 B003NRWVMC 1132 4.845406360424028 AAXUNK0W2DZG5 5.0 B003NRWVMC 1132 4.845406360424028 A237FS3MJZASVX 5.0 B003NRWVMC 1132 4.845406360424028 A2UQYI93VVN217 5.0 B003NRWVMC 1132 4.845406360424028 A1RYKEQH2466YY 5.0 B003NRWVMC 1132 4.845406360424028 A3Q6EZYHVUZ1YY 5.0 B003NRWVMC 1132 4.845406360424028 AG4WG8BBDJ08Y 5.0 B003NRWVMC 1132 4.845406360424028 A3MZSYME08MYMU 5.0 B003NRWVMC 1132 4.845406360424028 A2KOD57JZ2V17W 5.0 B003NRWVMC 1132 4.845406360424028 A1CUEPUYOH2ONP 5.0 B003NRWVMC 1132 4.845406360424028 A3JBBDCDXWPT4B 4.0 B003NRWVMC 1132 4.845406360424028 AEYA0U50498L6 5.0 B003NRWVMC 1132 4.845406360424028 AUWCUJULDL6AK 5.0 B003NRWVMC 1132 4.845406360424028 A3Q53VIDYRU8JD 4.0 B003NRWVMC 1132 4.845406360424028 A16YTPZBL682CG 5.0 B003NRWVMC 1132 4.845406360424028 A1U15HYI2HRW46 5.0 B003NRWVMC 1132 4.845406360424028 AC2KG4XYNQOXY 5.0 B003NRWVMC 1132 4.845406360424028 A3M9WHCEN1TNI3 5.0 B003NRWVMC 1132 4.845406360424028 A1F6BI2ZFS0AM6 5.0 B003NRWVMC 1132 4.845406360424028 A3EM3AAQCQ5MAU 5.0 B003NRWVMC 1132 4.845406360424028 ACLDY5BXMG1KL 5.0 B003NRWVMC 1132 4.845406360424028 A1JQXMDAK1DC0Q 5.0 B003NRWVMC 1132 4.845406360424028 A2GUATY34U52Y9 5.0 B003NRWVMC 1132 4.845406360424028 A3VLR6J5IYRTJF 5.0 B003NRWVMC 1132 4.845406360424028 A6ZAVMLXXUPQH 5.0 B003NRWVMC 1132 4.845406360424028 A1BC7EK0JZ7A8X 4.0 B003NRWVMC 1132 4.845406360424028 A3E36DNLOQPXHR 3.0 B003NRWVMC 1132 4.845406360424028 A3242P4VR1KZMJ 5.0 B003NRWVMC 1132 4.845406360424028 AEX6CPF0TE54Z 5.0 B003NRWVMC 1132 4.845406360424028 A3BPSEWY3XHH8N 4.0 B003NRWVMC 1132 4.845406360424028 A1HTKXKW1MG2K0 5.0 B003NRWVMC 1132 4.845406360424028 A19BC1SK27IJ2

#Prepare dataframe for pySpark

In order to build a model using pySpark ALS we have to remove the strings from the user and item variables since pySpark ALS does not take strings. The next step is to create corresponding numeric indecies for user and items. 
From Apache Spark Documentation
Note: The DataFrame-based API for ALS currently only supports integers for user and item ids. Other numeric types are supported for the user and item id columns, but the ids must be within the integer value range.

In [18]:
from pyspark.ml.feature import StringIndexer

In [19]:
stringIndexer = StringIndexer(inputCol="userId", outputCol="userIndex")
model = stringIndexer.fit(df)
dataIndexed = model.transform(df)

stringIndexer = StringIndexer(inputCol="itemId", outputCol="itemIndex")
model = stringIndexer.fit(dataIndexed)
dataIndexed = model.transform(dataIndexed)

display(dataIndexed)

userId rating itemId Count average userIndex itemIndex A3OSZISQRMTWZX 5.0 B003NRWVMC 1132 4.845406360424028 3930.0 60.0 AUMYZ3TS44MNP 5.0 B003NRWVMC 1132 4.845406360424028 31909.0 60.0 A3RA3ZTC0HDP9U 5.0 B003NRWVMC 1132 4.845406360424028 15727.0 60.0 A18NIX1QH8L1Z9 4.0 B003NRWVMC 1132 4.845406360424028 85495.0 60.0 A1HQ4RXXJNS5TC 5.0 B003NRWVMC 1132 4.845406360424028 114301.0 60.0 A36GS5K3YYXFMX 5.0 B003NRWVMC 1132 4.845406360424028 148614.0 60.0 A34IPQQL7NIJQL 5.0 B003NRWVMC 1132 4.845406360424028 65610.0 60.0 A20ZO8UFW5EFS6 5.0 B003NRWVMC 1132 4.845406360424028 6588.0 60.0 A38IOP4TW32Y2Q 5.0 B003NRWVMC 1132 4.845406360424028 48178.0 60.0 A133G8UNAV44BP 5.0 B003NRWVMC 1132 4.845406360424028 6488.0 60.0 AEKZL04PZT00E 5.0 B003NRWVMC 1132 4.845406360424028 98198.0 60.0 A2ZTOTUEBOOFQO 4.0 B003NRWVMC 1132 4.845406360424028 68385.0 60.0 A1ZN17SY8B3LCZ 5.0 B003NRWVMC 1132 4.845406360424028 15333.0 60.0 A3P6L56GUQ85Q2 5.0 B003NRWVMC 1132 4.845406360424028 61458.0 60.0 A1AUP6RUJD01IW 5.0 B003NRWVMC 1132 4.845406360424028 41994.0 60.0 A1S7DJPSFDSEIW 5.0 B003NRWVMC 1132 4.845406360424028 45869.0 60.0 AFTCML5ZKDJ2F 5.0 B003NRWVMC 1132 4.845406360424028 97242.0 60.0 A1EUIHORKSTT3T 5.0 B003NRWVMC 1132 4.845406360424028 17691.0 60.0 AP3JFZYTXTD68 5.0 B003NRWVMC 1132 4.845406360424028 133950.0 60.0 AT0CF2RA9EZHG 5.0 B003NRWVMC 1132 4.845406360424028 3806.0 60.0 A1M0OOSTWL760H 5.0 B003NRWVMC 1132 4.845406360424028 13405.0 60.0 A11J7EOZFK4VX0 5.0 B003NRWVMC 1132 4.845406360424028 143551.0 60.0 A3MICL5RG8QZX7 5.0 B003NRWVMC 1132 4.845406360424028 73525.0 60.0 A2CUZL460CVQP5 5.0 B003NRWVMC 1132 4.845406360424028 147672.0 60.0 A1GP6FF2KAJ4OQ 5.0 B003NRWVMC 1132 4.845406360424028 88070.0 60.0 A4YOGHJ318UGY 4.0 B003NRWVMC 1132 4.845406360424028 48197.0 60.0 AYJKFTHNF6BYD 5.0 B003NRWVMC 1132 4.845406360424028 9639.0 60.0 A2MYD85NGIUWE6 5.0 B003NRWVMC 1132 4.845406360424028 106558.0 60.0 APT5R9GEC52L8 5.0 B003NRWVMC 1132 4.845406360424028 164837.0 60.0 A5TDOCTCGY19V 5.0 B003NRWVMC 1132 4.845406360424028 8405.0 60.0 A3NT2KWVZ8WEBJ 5.0 B003NRWVMC 1132 4.845406360424028 51438.0 60.0 A2QYY2GU66WDXU 5.0 B003NRWVMC 1132 4.845406360424028 7874.0 60.0 AGOYTDOB23CNY 5.0 B003NRWVMC 1132 4.845406360424028 136932.0 60.0 A1C7AQVHUDWFY4 4.0 B003NRWVMC 1132 4.845406360424028 15950.0 60.0 ASV9ERSOWJN5V 5.0 B003NRWVMC 1132 4.845406360424028 74522.0 60.0 A2G10P9DQS555B 5.0 B003NRWVMC 1132 4.845406360424028 9439.0 60.0 ALE8NCB09O7PG 5.0 B003NRWVMC 1132 4.845406360424028 149633.0 60.0 A193F6OTQGTWFJ 5.0 B003NRWVMC 1132 4.845406360424028 19170.0 60.0 A2VDQ2WIVGFNFD 5.0 B003NRWVMC 1132 4.845406360424028 107171.0 60.0 A3MEUXPK3NVFK7 5.0 B003NRWVMC 1132 4.845406360424028 49399.0 60.0 A1LJK5FVKFDXZU 4.0 B003NRWVMC 1132 4.845406360424028 87154.0 60.0 AIDK7FMI74XI 5.0 B003NRWVMC 1132 4.845406360424028 4978.0 60.0 AWXT84BCMQ5QK 5.0 B003NRWVMC 1132 4.845406360424028 12901.0 60.0 A1U58K0US2NHWV 5.0 B003NRWVMC 1132 4.845406360424028 139950.0 60.0 A2AVVEH1V53IJW 5.0 B003NRWVMC 1132 4.845406360424028 101028.0 60.0 AAXUNK0W2DZG5 5.0 B003NRWVMC 1132 4.845406360424028 39882.0 60.0 A237FS3MJZASVX 5.0 B003NRWVMC 1132 4.845406360424028 36382.0 60.0 A2UQYI93VVN217 5.0 B003NRWVMC 1132 4.845406360424028 154587.0 60.0 A1RYKEQH2466YY 5.0 B003NRWVMC 1132 4.845406360424028 54368.0 60.0 A3Q6EZYHVUZ1YY 5.0 B003NRWVMC 1132 4.845406360424028 127465.0 60.0 AG4WG8BBDJ08Y 5.0 B003NRWVMC 1132 4.845406360424028 163332.0 60.0 A3MZSYME08MYMU 5.0 B003NRWVMC 1132 4.845406360424028 111510.0 60.0 A2KOD57JZ2V17W 5.0 B003NRWVMC 1132 4.845406360424028 13604.0 60.0 A1CUEPUYOH2ONP 5.0 B003NRWVMC 1132 4.845406360424028 109508.0 60.0 A3JBBDCDXWPT4B 4.0 B003NRWVMC 1132 4.845406360424028 73761.0 60.0 AEYA0U50498L6 5.0 B003NRWVMC 1132 4.845406360424028 136118.0 60.0 AUWCUJULDL6AK 5.0 B003NRWVMC 1132 4.845406360424028 79896.0 60.0 A3Q53VIDYRU8JD 4.0 B003NRWVMC 1132 4.845406360424028 30565.0 60.0 A16YTPZBL682CG 5.0 B003NRWVMC 1132 4.845406360424028 1100.0 60.0 A1U15HYI2HRW46 5.0 B003NRWVMC 1132 4.845406360424028 121016.0 

In [20]:
dataIndexed.printSchema()

root
-- userId: string (nullable = true)
-- rating: string (nullable = true)
-- itemId: string (nullable = true)
-- Count: long (nullable = false)
-- average: double (nullable = true)
-- userIndex: double (nullable = false)
-- itemIndex: double (nullable = false)

Now we have userIndex, itemIndex corresponding with userId and ItemId and we are going to select only the columns required for modeling and cast rating as a float

In [22]:
df=dataIndexed.select(dataIndexed.userId,dataIndexed.itemId,dataIndexed.rating.cast("float"),dataIndexed.userIndex,dataIndexed.itemIndex)
df.printSchema()

root
-- userId: string (nullable = true)
-- itemId: string (nullable = true)
-- rating: float (nullable = true)
-- userIndex: double (nullable = false)
-- itemIndex: double (nullable = false)

Split data .8 train and .2 test

In [24]:
(training,test)=df.randomSplit([0.8,0.2])

print('Training: {0}, test: {1}\n'.format(
  training.count(), test.count())
)
training.show(3)
test.show(3)

Training: 154056, test: 38321

+--------------------+----------+------+---------+---------+
 userId| itemId|rating|userIndex|itemIndex|
+--------------------+----------+------+---------+---------+
A004476834K8BM0W3...|B000MVN8GK| 5.0| 94042.0| 51.0|
A03373291U5ZIN1CQ...|B000MVN8GK| 4.0| 107488.0| 51.0|
A04679053BE8ZYGJ7...|B000VZUWZW| 4.0| 42554.0| 56.0|
+--------------------+----------+------+---------+---------+
only showing top 3 rows

+--------------+----------+------+---------+---------+
 userId| itemId|rating|userIndex|itemIndex|
+--------------+----------+------+---------+---------+
A100SBTKAMHC8R|B001F6ZIXC| 5.0| 120378.0| 30.0|
A103JQ41DH2YU9|B000MVN8HE| 1.0| 156114.0| 32.0|
A106WN5A9FJJGQ|B000MVN8HE| 4.0| 123823.0| 32.0|
+--------------+----------+------+---------+---------+
only showing top 3 rows

#Model

Train ALS models with different K

In [26]:
from pyspark.ml.recommendation import ALS
from pyspark.ml.tuning import TrainValidationSplit,ParamGridBuilder
from pyspark.ml.evaluation import RegressionEvaluator

"ALS is an iterative optimization process where we for every iteration try to arrive closer and closer to a factorized representation of our original data."
Just like other recommender systems ALS has it's own hyper-parameters that need to be tuned and for that purpose we are using the build-in function ParamGridBuilder to come up with the best combination of:

•	rank: the number of latent factors in the model (defaults to 10)
•	maxIter: the maximum number of iterations to run (defaults to 10)
•	regParam: the regularization parameter in ALS (defaults to 1.0)

In [28]:
als = ALS()

als.setUserCol('userIndex')\
   .setItemCol('itemIndex')

param_grid=ParamGridBuilder()\
            .addGrid(als.rank,[8,12,20])\
            .addGrid(als.maxIter,[10,20])\
            .addGrid(als.regParam,[.05, .1])\
            .build()

evaluator=RegressionEvaluator(metricName="rmse",labelCol="rating",predictionCol="prediction")

tvs=TrainValidationSplit(estimator=als,estimatorParamMaps=param_grid,evaluator=evaluator)

In [29]:
# Import mlflow
import mlflow
import mlflow.sklearn

Mon Apr 20 12:33:02 2020 py4j imported

In [30]:
model=tvs.fit(training)

java.io.IOException: Connection failed
	at com.databricks.rpc.Jetty9Client$$anon$1.handleError(Jetty9Client.scala:468)
	at com.databricks.rpc.Jetty9Client$$anon$1.onFailure(Jetty9Client.scala:395)
	at shaded.v9_4.org.eclipse.jetty.client.ResponseNotifier.notifyFailure(ResponseNotifier.java:177)
	at shaded.v9_4.org.eclipse.jetty.client.ResponseNotifier.notifyFailure(ResponseNotifier.java:169)
	at shaded.v9_4.org.eclipse.jetty.client.HttpExchange.notifyFailureComplete(HttpExchange.java:268)
	at shaded.v9_4.org.eclipse.jetty.client.HttpExchange.abort(HttpExchange.java:240)
	at shaded.v9_4.org.eclipse.jetty.client.HttpConversation.abort(HttpConversation.java:149)
	at shaded.v9_4.org.eclipse.jetty.client.HttpRequest.abort(HttpRequest.java:768)
	at shaded.v9_4.org.eclipse.jetty.client.HttpDestination.abort(HttpDestination.java:453)
	at shaded.v9_4.org.eclipse.jetty.client.HttpDestination.failed(HttpDestination.java:234)
	at shaded.v9_4.org.eclipse.jetty.client.AbstractConnectionPool$1.failed(AbstractConnectionPool.java:140)
	at shaded.v9_4.org.eclipse.jetty.util.Promise$Wrapper.failed(Promise.java:136)
	at shaded.v9_4.org.eclipse.jetty.client.HttpClient$1$1.failed(HttpClient.java:612)
	at shaded.v9_4.org.eclipse.jetty.client.AbstractConnectorHttpClientTransport.connectFailed(AbstractConnectorHttpClientTransport.java:138)
	at shaded.v9_4.org.eclipse.jetty.client.AbstractConnectorHttpClientTransport$ClientSelectorManager.connectionFailed(AbstractConnectorHttpClientTransport.java:188)
	at shaded.v9_4.org.eclipse.jetty.io.ManagedSelector$Connect.failed(ManagedSelector.java:822)
	at shaded.v9_4.org.eclipse.jetty.io.ManagedSelector.processConnect(ManagedSelector.java:254)
	at shaded.v9_4.org.eclipse.jetty.io.ManagedSelector.access$1400(ManagedSelector.java:62)
	at shaded.v9_4.org.eclipse.jetty.io.ManagedSelector$SelectorProducer.processSelected(ManagedSelector.java:543)
	at shaded.v9_4.org.eclipse.jetty.io.ManagedSelector$SelectorProducer.produce(ManagedSelector.java:401)
	at shaded.v9_4.org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produceTask(EatWhatYouKill.java:360)
	at shaded.v9_4.org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:184)
	at shaded.v9_4.org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
	at shaded.v9_4.org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129)
	at shaded.v9_4.org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:367)
	at shaded.v9_4.org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:782)
	at shaded.v9_4.org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:914)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.net.ConnectException: Connection refused
	at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
	at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:714)
	at shaded.v9_4.org.eclipse.jetty.io.SelectorManager.doFinishConnect(SelectorManager.java:355)
	at shaded.v9_4.org.eclipse.jetty.io.ManagedSelector.processConnect(ManagedSelector.java:232)
	... 11 more

In [31]:
best_model=model.bestModel

predictions=best_model.transform(test)

predicted_ratings_df = predictions.filter(predictions.prediction != float('nan'))

rmse=evaluator.evaluate(predicted_ratings_df)

print("RMSE = " + str(rmse))
print("**Best Model**")
print(" Rank:"),best_model.rank
print(" MaxIter:"),best_model._java_obj.parent().getMaxIter()
print(" RegParam:"),best_model._java_obj.parent().getRegParam()

The best model has 22 latent factors and regulization = 0.1
The RMSE is 1.89644273379, which is bigger than both IBCF= 1.1570069 and UBCF= 0.9240670 from the previous project

In [33]:
display(predicted_ratings_df)

userId itemId rating userIndex itemIndex prediction A3UZV6KHEPQKQ5 B00CDBQNZU 5.0 17753.0 31.0 2.8665416 A2UWXEEWIKFPBP B00CDBQNZU 5.0 897.0 31.0 3.7309794 A2Y2JTGER56QR4 B00CDBQNZU 5.0 4119.0 31.0 1.6496632 A2L3FAPDVEUL5K B00CDBQNZU 1.0 16234.0 31.0 2.8665416 A07210292XA7ZPKESLMC8 B00CDBQNZU 4.0 5375.0 31.0 2.9191685 A2T4OL4DO5CF6V B00CDBQNZU 5.0 15093.0 31.0 3.5831769 A3S48Q63UPO9L6 B00CDBQNZU 5.0 13170.0 31.0 2.323049 AEQ5O09E7ERTB B00CDBQNZU 4.0 18193.0 31.0 -0.09269881 A2NJOS6PYGAB5H B00CDBQNZU 5.0 17196.0 31.0 2.199551 A1CFRFLS49G8R8 B00CDBQNZU 1.0 3313.0 31.0 3.799327 A22GWMIVVT70FD B00CDBQNZU 5.0 9217.0 31.0 3.5831769 A1L7VHZBTOMQ9L B00CDBQNZU 5.0 1031.0 31.0 3.210308 A3QZ3XTC2FKTF8 B00CDBQNZU 5.0 3264.0 31.0 2.6968527 AHPGDBGUJ7BQM B00CDBQNZU 5.0 9147.0 31.0 3.5831769 A1Y5S2K97UXRVO B00CDBQNZU 5.0 15905.0 31.0 0.7166354 A1FGOX1XHXHYY4 B00CDBQNZU 5.0 3054.0 31.0 3.334064 AXCD45O7GCNLT B00CDBQNZU 5.0 17285.0 31.0 3.5831769 AIUICWDPQL2BE B00CDBQNZU 5.0 15401.0 31.0 2.3353348 A1M5LJSCJ19S6F B00CDBQNZU 5.0 663.0 31.0 2.6833773 A3TXW0UJQHGIIZ B00CDBQNZU 5.0 2774.0 31.0 3.334064 A2FO61JYXOKX5M B00CDBQNZU 5.0 16285.0 31.0 2.8665416 A1GHSBDNMOL6K8 B00CDBQNZU 5.0 4784.0 31.0 3.5831769 A3VI36BFNGUKGG B00CDBQNZU 3.0 14656.0 31.0 0.5498878 ANU0FJEZ4L3WM B00CDBQNZU 5.0 18128.0 31.0 0.5498878 A2XG6VCDBYCMLV B00CDBQNZU 4.0 4469.0 31.0 1.6758901 A257RVFKMH64Y8 B00CDBQNZU 2.0 1145.0 31.0 2.2272873 A24IVM7M7PTUE9 B00CDBQNZU 5.0 1304.0 31.0 1.101097 AQ91QTWJ2RORA B00CDBQNZU 4.0 2786.0 31.0 3.6300392 A2S8J9H2JUY66M B00CDBQNZU 5.0 16811.0 31.0 0.27557707 ALQQIHJI4937D B00CDBQNZU 5.0 2685.0 31.0 2.211674 A2EBY3BNJ8KI52 B00CDBQNZU 5.0 11750.0 31.0 2.199551 AWCHKM1P8RL6H B00CDBQNZU 2.0 925.0 31.0 0.36579984 A1KEB6YHR5CUYD B00CDBQNZU 5.0 10323.0 31.0 2.9191685 A2XR7HUK0PCVXI B00CDBQNZU 5.0 3415.0 31.0 2.1499062 A2P3OXEDV6DXFY B00CDBQNZU 3.0 6775.0 31.0 0.68894255 A2GIOSNLABX3JM B00CDBQNZU 5.0 5010.0 31.0 1.0997756 A1BUCCYTMYFIK2 B00CDBQNZU 4.0 8912.0 31.0 1.3737168 A2F16QVMHMO6YN B00CDBQNZU 1.0 12353.0 31.0 2.7494388 ACPFU2KWSWTH4 B00CDBQNZU 5.0 14399.0 31.0 1.005904 A2LP63NI1A3ANH B00CDBQNZU 4.0 1222.0 31.0 2.4455893 A3LZA698SQPCXE B00CDBQNZU 5.0 4247.0 31.0 3.7309794 A3I2RAG9Z8EZY7 B00CDBQNZU 5.0 6646.0 31.0 3.5831769 A2A2UF2O5F6PSL B00CDBQNZU 5.0 15336.0 31.0 2.7494388 A34JO9MTFD353D B00CDBQNZU 5.0 4448.0 31.0 2.7494388 A1SB9DPBTPJMD B00CDBQNZU 1.0 2764.0 31.0 2.4314044 AGZ2EYS1HA4N2 B00CDBQNZU 5.0 2324.0 31.0 2.7348492 ASFUBTES6349U B00CDBQNZU 5.0 2445.0 31.0 2.932419 A3MBV8Y5I2QWED B00CDBQNZU 3.0 12655.0 31.0 -1.8969357 A3N6G5OH2HX309 B00CDBQNZU 5.0 8202.0 31.0 2.9191685 A1H944KVMHQIFP B00CDBQNZU 5.0 17428.0 31.0 3.5831769 A2K8Y8V2X7KJSY B00CDBQNZU 5.0 2988.0 31.0 1.4332708 A341LQOPI7GC5U B00CDBQNZU 3.0 9198.0 31.0 3.5831769 A2QB3BB7OEQIB3 B00CDBQNZU 5.0 957.0 31.0 1.6496632 A37XMVKGIYZQOA B00CDBQNZU 4.0 1381.0 31.0 1.4592688 A161OPG1GEMU83 B00CDBQNZU 4.0 1910.0 31.0 1.6496632 A24JXAHCM04JR1 B00CDBQNZU 5.0 17670.0 31.0 1.0997756 AGQUNWKXUXT99 B00CDBQNZU 4.0 11816.0 31.0 1.7515008 A1B4LLFMZN4A8T B00CDBQNZU 4.0 11576.0 31.0 2.4796226 A3UK9T99SYY3EU B00CDBQNZU 4.0 16106.0 31.0 2.199551 A3EPSUZPMQ8FDE B00CDBQNZU 4.0 3851.0 31.0 2.199551 AATOOVSQ0S8M9 B00CDBQNZU 2.0 18592.0 31.0 3.5831769 A2FQV1DD7QT67A B00CDBQNZU 4.0 5475.0 31.0 1.6496632 A35DOM42O00RZE B00CDBQNZU 3.0 1409.0 31.0 2.4678788 AAZ66E4DHPQ3I B00CDBQNZU 5.0 712.0 31.0 2.8092816 AJYE6BP33MV0 B00CDBQNZU 1.0 11220.0 31.0 2.8665416 A2MH054G8CRIDZ B00CDBQNZU 5.0 953.0 31.0 2.9958014 A1I9MSKCK9F4C9 B00CDBQNZU 5.0 1440.0 31.0 2.8157554 A161I8TNQWW1UL B00CDBQNZU 5.0 14487.0 31.0 0.39798328 A3H3RHSUWQK5J6 B00CDBQNZU 5.0 2058.0 31.0 0.86607265 ASIF3PZJYQAGZ B00CDBQNZU 4.0 10784.0 31.0 1.3484298 A2H847LNV87AA0 B00CDBQNZU 4.0 6023.0 31.0 2.7494388 A244XA5SXYSKBF B00CDBQNZU 3.0 2486.0 31.0 2.9847834 A1AYDULRCAF3IF B00CDBQNZU 5.0 11634.0 31.0 2.8665416 A3MGN83A36KEMK B00CDBQNZU 5.0 1992.0 31.0 1.0955584 A2815UMO47O55J B00CDBQNZU 4.0 14911.0 31.0 2.199551 AO0A7FJH56H17 B00CDBQNZU 3.0 1004.0 31.0 

In [34]:
reg_eval = RegressionEvaluator(predictionCol="prediction", labelCol="rating", metricName="rmse")

#Recommendations

Generate Recommendations for all users

In [36]:
user_rec=best_model.recommendForAllUsers(10)

In [37]:
display(user_rec)

userIndex recommendations 148 List(List(15, 4.9310813), List(46, 4.7800746), List(6, 4.7537084), List(63, 4.7201915), List(1, 4.637907), List(60, 4.601981), List(10, 4.5879793), List(18, 4.5705414), List(39, 4.5512996), List(20, 4.523264)) 463 List(List(24, 6.560636), List(62, 6.3836765), List(3, 6.249261), List(25, 6.135927), List(61, 5.940811), List(21, 5.9244184), List(53, 5.7974215), List(64, 5.7418127), List(57, 5.699632), List(33, 5.609172)) 471 List(List(62, 6.1830344), List(32, 4.943763), List(7, 4.702999), List(14, 4.658553), List(61, 4.318445), List(12, 4.3122706), List(65, 3.9326615), List(1, 3.5566816), List(3, 3.4101014), List(37, 3.2326102)) 496 List(List(12, 5.872194), List(32, 5.850674), List(51, 5.5642624), List(37, 5.533883), List(61, 5.507338), List(29, 5.4738364), List(9, 5.361377), List(18, 5.3560534), List(23, 5.082713), List(14, 5.0502806)) 833 List(List(62, 6.6936264), List(3, 4.775024), List(31, 4.4702916), List(17, 4.4406714), List(36, 4.314538), List(61, 4.2573624), List(1, 4.2495894), List(65, 4.0683), List(30, 3.9462829), List(38, 3.9390357)) 1088 List(List(54, 5.4669986), List(28, 5.211794), List(26, 5.1970687), List(4, 5.1481004), List(68, 5.0985236), List(35, 5.04649), List(27, 4.960038), List(16, 4.930458), List(5, 4.920898), List(56, 4.769057)) 1238 List(List(67, 5.0436606), List(64, 5.037113), List(6, 5.035599), List(33, 4.958264), List(29, 4.907932), List(39, 4.8235645), List(55, 4.818003), List(35, 4.756812), List(28, 4.751462), List(15, 4.7511744)) 1342 List(List(15, 4.9602337), List(10, 4.8338776), List(18, 4.826921), List(60, 4.802887), List(63, 4.757739), List(46, 4.611818), List(6, 4.576256), List(59, 4.5233107), List(58, 4.4695516), List(2, 4.463224)) 1580 List(List(28, 5.0033064), List(67, 4.973325), List(0, 4.8970094), List(22, 4.8352075), List(11, 4.8274565), List(8, 4.772457), List(68, 4.7621307), List(55, 4.752386), List(35, 4.7518287), List(49, 4.701833)) 1591 List(List(24, 5.280665), List(54, 5.0922127), List(56, 4.6291947), List(41, 4.614948), List(50, 4.5113573), List(44, 4.4997964), List(43, 4.4841704), List(0, 4.3178077), List(28, 4.2971864), List(27, 4.229585)) 1829 List(List(11, 5.0886264), List(8, 5.0604353), List(22, 4.878569), List(49, 4.87597), List(39, 4.6585207), List(0, 4.627097), List(36, 4.621004), List(67, 4.605047), List(25, 4.6006746), List(4, 4.567636)) 1959 List(List(67, 5.221072), List(35, 5.0725207), List(28, 5.065014), List(55, 5.056529), List(68, 5.0361943), List(6, 4.9707203), List(64, 4.942608), List(16, 4.9259253), List(33, 4.9018083), List(15, 4.8649936)) 2122 List(List(62, 5.795812), List(7, 3.7597957), List(3, 3.4036875), List(17, 3.2168465), List(61, 3.1030226), List(21, 2.9358108), List(31, 2.9020176), List(32, 2.8644583), List(65, 2.7034936), List(12, 2.6686735)) 2142 List(List(12, 4.777539), List(32, 4.5550385), List(37, 4.3183594), List(51, 4.277199), List(61, 4.247382), List(62, 4.2428308), List(24, 4.153714), List(29, 4.119118), List(9, 3.9980693), List(14, 3.9131806)) 2366 List(List(10, 3.5789392), List(26, 3.4985526), List(68, 3.3045733), List(60, 3.262006), List(2, 3.2364914), List(35, 3.2288592), List(16, 3.2203438), List(59, 3.2184129), List(27, 3.2183084), List(15, 3.1537519)) 2659 List(List(62, 5.9421396), List(17, 4.241562), List(31, 4.119923), List(30, 3.53367), List(3, 3.2544212), List(34, 3.121674), List(36, 2.9633756), List(7, 2.9419758), List(1, 2.5355077), List(38, 2.3884785)) 2866 List(List(12, 2.8868399), List(24, 2.6828566), List(32, 2.6569912), List(37, 2.6243353), List(51, 2.5996747), List(61, 2.5059967), List(29, 2.4821553), List(62, 2.4727979), List(9, 2.4049888), List(23, 2.2848394)) 3175 List(List(62, 4.6746902), List(24, 4.17572), List(21, 3.8392665), List(3, 3.7960067), List(25, 3.5621376), List(57, 3.4638596), List(53, 3.4453697), List(66, 3.1571789), List(61, 3.1023872), List(7, 3.100457)) 3749 List(List(10, 4.8833275), List(63, 4.2941127), List(60, 4.2610316), List(46, 4.236853), List(15, 4.21826), List(59, 4.0629

In [38]:
user_rec.count()

Out[ 85 ]: 136375

Produce recommendations for one user

In [40]:
user = user_rec.filter(user_rec.userIndex==148)

In [41]:
display(user)

userIndex recommendations 148 List(List(15, 4.9310813), List(46, 4.7800746), List(6, 4.7537084), List(63, 4.7201915), List(1, 4.637907), List(60, 4.601981), List(10, 4.5879793), List(18, 4.5705414), List(39, 4.5512996), List(20, 4.523264))

We can see that the algorithm is recommnding items with similar ratings, which is an indication that a user is getting similar items recommnded to them

In [43]:
rec=user.select("recommendations.itemIndex","recommendations.rating")

In [44]:
display(rec)

itemIndex rating List(15, 46, 6, 63, 1, 60, 10, 18, 39, 20) List(4.9310813, 4.7800746, 4.7537084, 4.7201915, 4.637907, 4.601981, 4.5879793, 4.5705414, 4.5512996, 4.523264)

In [45]:
import pandas as pd

In [46]:
rec_items=rec.select("itemIndex").toPandas().iloc[0,0]
rec_ratings=rec.select("rating").toPandas().iloc[0,0]

In [47]:
rec_matrix=pd.DataFrame(rec_items,columns=["itemsIndex"])
rec_matrix["ratings"]=rec_ratings
ratings_matrix_df=sqlContext.createDataFrame(rec_matrix)
display(ratings_matrix_df)

itemsIndex ratings 15 4.931081295013428 46 4.780074596405029 6 4.753708362579346 63 4.720191478729248 1 4.637907028198242 60 4.601981163024902 10 4.587979316711426 18 4.5705413818359375 39 4.551299571990967 20 4.523263931274414

We map the index of the items recommended back to their ids

In [49]:
test_items=test.dropDuplicates(['itemId','itemIndex']).select("itemId","itemIndex")
display(test_items)

itemId itemIndex B004GTOKM0 67.0 B00BLCHYKU 38.0 B0083IJGBU 45.0 B000VZUWZW 56.0 B006Z48TZS 15.0 B00C7KXUOE 20.0 B00L86ZKAK 61.0 B00I3MMN4I 2.0 B003RRW3BC 58.0 B008EQHT4M 54.0 B00CDBR1P6 7.0 B000MVN8HE 32.0 B008ZXSUO2 44.0 B00DKRXM7S 53.0 B00C0OLT6S 33.0 B00CDBTQCW 3.0 B00DQISQX6 37.0 B00F406S2U 57.0 B004MWZLYC 18.0 B00I3MPDP4 1.0 B00I3MNVBW 14.0 B00I3MMLHC 65.0 B009SJ3GUK 22.0 B00CDBQNZU 31.0 B007JF83YY 13.0 B00F3WB7FM 25.0 B0012QRPU4 47.0 B0040HNZTW 29.0 B009FZFONO 5.0 B00CE10BD4 64.0 B00AX4QEAS 66.0 B00I3MMQYK 17.0 B00BLCHL4Y 16.0 B003B63FKW 40.0 B009JARQ3G 35.0 B00B8P8O9K 6.0 B00D1N6OQ2 68.0 B007QYRCX6 62.0 B0099JKR6U 11.0 B00I3MMTS8 21.0 B001F6ZIXC 30.0 B00BW75MKI 48.0 B008BQG3RE 27.0 B009IF7092 55.0 B008BQ8VOC 23.0 B00H7NDSPC 46.0 B00932ADLC 52.0 B009IJJ5B4 59.0 B009IG59DK 49.0 B00FMO8DXU 36.0 B00D5OZQUC 26.0 B00DTOYIIE 4.0 B008QTTGGG 19.0 B005544TRQ 9.0 B009AP3HQK 42.0 B00CDBTVM2 24.0 B000OGTRC2 43.0 B00CBNOD0W 34.0 B00F0XPJH6 39.0 B003NRWVMC 60.0 B003NRVE6Q 63.0 B000OJFV3S 41.0 B00DAHSVYC 0.0 B00F87ZUYG 8.0 B008JSO21S 28.0 B00I3MNGCG 12.0 B000MVN8GK 51.0 B00APE00H4 10.0 B004KPX56C 50.0

In [50]:
user_recommendations=test_items.join(ratings_matrix_df,test_items.itemIndex==ratings_matrix_df.itemsIndex).drop("itemIndex","itemsIndex")
display(user_recommendations)

itemId ratings B004MWZLYC 4.5705413818359375 B00I3MPDP4 4.637907028198242 B00F0XPJH6 4.551299571990967 B003NRVE6Q 4.720191478729248 B003NRWVMC 4.601981163024902 B00H7NDSPC 4.780074596405029 B00APE00H4 4.587979316711426 B00B8P8O9K 4.753708362579346 B00C7KXUOE 4.523263931274414 B006Z48TZS 4.931081295013428

This table shows the recommended items for the selected user and the corresponding ratings.

##Conclusion

The goal of this project was to practice working with a distributed recommnder system. 
We were able to adapt an ALS recommender system to work with Apache Spark via Databricks platform. The dataset that was used to build the model was the same as in Project 4 where IBCF anf UBCF collaborative systems were build in order to recommend new items to users. The best performing algorithm was UBCF with RMSE of 0.9240670. 
The advantages of working with Spark as oppose to local memory were its ability to handle the entire dataset of 583933 where before was almost impossible and the high-level libraries, including support for SQL queries.

The architechture of Apache Spark-- "At a fundamental level, an Apache Spark application consists of two main components: a driver, which converts the user's code into multiple tasks that can be distributed across worker nodes, and executors, which run on those nodes and execute the tasks assigned to them. Some form of cluster manager is necessary to mediate between the two." from https://www.infoworld.com/article/3236869/what-is-apache-spark-the-big-data-platform-that-crushed-hadoop.html

The ALS model that we build resulted in 22 latent factors and regulization = 0.1 with RMSE of 1.89644273379, which is bigger than both IBCF= 1.1570069 and UBCF= 0.9240670 from the previous project. Although when evaluating the model we can see that the items that were recommended to one user were all with similar ratings.
A possible improvemnet of the model might include some implicit data where we can also have additional information such as genre, number of views, clicks, purchases etc and set the implicitPrefs=True to get a better idea of the similarity of the iteams and also to improve the results/improve the model.